Implement NCCH packing

WorstAquaPlayer opened the pull request:
### Description

Implement converter to pack a Ncch class into a binary NCCH format.
Firstly, I deeply apologize if some of my code is pure aberration, I'm still learning how to use Yarhl (and how to properly code in general).
To be able to pack a Ncch back to a binary format, it now reads all the necessary header bytes that were not implemented when reading a Ncch binary.
The updated integration test generates a binary that is equal to the original one. (This was only tested with the games Phoenix Wright Ace Attorney Trilogy and Phoenix Wright Ace Attorney Dual Destinies)
One thing should be noted, generating a NCCH that has been modified may or may not work on a generated CIA, since the TMD table isn't updated with this new data.

### Example

```cs
ncch.TransformWith<Ncch2Binary>();
```

This closes #8

WorstAquaPlayer committed:
:sparkles: Implement NCCH packing

WorstAquaPlayer committed:
:umbrella: Add NCCH pack test
Also changed the unpack test to behave similarly to the NodeContainer tests, while using the original methods.

WorstAquaPlayer committed:
Read and use later the region size
It was always asumed that 1 was the value for both System and Rom hashes size.

WorstAquaPlayer committed:
Fix wrong size and hash in the fake NCCH

WorstAquaPlayer committed:
NCCH pack child in the same offsets (if possible)
Now, the children are not arbitrarily written in the end of the stream. With this, tests of retail games should generate a binary that is equal to the original one.

WorstAquaPlayer committed:
Properly find if the chunk was used or not.
Previously, it would only search if the offset was in the used range, but didn't mind the length, so it would write over the previous content.

pleonex commented on src/Lemon/Containers/Converters/Ncch2Binary.cs:
```diff
@@ -0,0 +1,237 @@
+
+            // Write the full size of the binary (in units) in 0x104
+            writer.Stream.Position = 0x104;
+            writer.Write((int)writer.Stream.Length / NcchHeader.Unit);
```
I think that if the division is not exact, then we will need to add padding and write a higher number. For instance, if Stream is 768 bytes, we need to write `2` but right now it would be `768 / 512 = 1`.
We can get this done by simply calling before this line: `writer.WritePadding(0, NcchHeader.Unit);`

It will do nothing if the length is already a multiple of 512 (Unit), otherwise it will write padding and then we will get the correct stream length.

pleonex commented on src/Lemon/Containers/Converters/Ncch2Binary.cs:
```diff
@@ -0,0 +1,237 @@
+                        writer.Write(sha256.ComputeHash(file.Stream));
+                    }
+                } catch (Exception ex) {
+                    throw new FormatException(ex.Message);
```
To encapsulate an exception it's better to include the full `ex` variable, so we can access to its information (like stacktrace and line numbers) via the `InnerException` property of this new format exception. Something like:
```csharp
throw new FormatException($"Failed to perform SHA256 for {file.Name}", ex);
```

pleonex commented on src/Lemon/Containers/Converters/Ncch2Binary.cs:
```diff
@@ -0,0 +1,237 @@
+                long position = writer.Stream.Position;
+                int offset = offsetInUnits * NcchHeader.Unit;
+                int fileLength = (int)file.Stream.Length;
+                int fileLengthInUnits = fileLength / NcchHeader.Unit;
```
As it happened with the NCCH length, before dividing we need to pad. We could do it here like `fileLength.Pad(NcchHeader.Unit) / NcchHeader.Unit` or after writing the file and its padding, do the division.

pleonex commented on src/Lemon/Containers/Converters/Ncch2Binary.cs:
```diff
@@ -0,0 +1,237 @@
+
+                bool usedChunk = false;
+                var fileRange = Enumerable.Range(offset, fileLength);
+                foreach (var usedUnit in usedUnits) {
```
What's the goal of this dictionary?

WorstAquaPlayer replied:
This dictionary contains the starting offset and the length of already written chunks, so later it can be used to determine before writing a new chunk if there was content in that range. This has some relation to a later comment, this is done to generate an equal stream to the original one while testing.

pleonex commented on src/Lemon/Containers/Converters/Ncch2Binary.cs:
```diff
@@ -0,0 +1,237 @@
+                    }
+                }
+
+                if (writer.Stream.Length > offset && usedChunk) {
+                    offset = (int)Math.Ceiling((double)writer.Stream.Length / NcchHeader.Unit) * NcchHeader.Unit;
+                    offsetInUnits = offset / NcchHeader.Unit;
+                }
+
+                if (writer.Stream.Length < offset) {
+                    long tempPosition = writer.Stream.Position;
+
+                    writer.Stream.Position = writer.Stream.Length;
+                    writer.Write(new byte[offset - writer.Stream.Length]);
+
+                    writer.Stream.Position = tempPosition;
+                }
```
I think we needed to do this because we are not writing the padding after each file. I think we won't needed once you add the line from the comment below.

WorstAquaPlayer replied:
That's partially the reason, but it's also done for generating the same stream while testing and if the content has changed then it calculates a new position at the end of the file. But yeah, it could use some padding.

pleonex commented on src/Lemon/Containers/Converters/Ncch2Binary.cs:
```diff
@@ -0,0 +1,237 @@
+
+                writer.Stream.Position = offset;
+
+                file.Stream.WriteTo(writer.Stream);
```
After writing the file, we need to add the padding to the unit: `writer.WritePadding(0, NcchHeader.Unit);`

pleonex replied:
The file padding is still missing, right?

WorstAquaPlayer replied:
I don't know how this works, but this isn't the content of the latest commit, in there the file length is padded, so I think that should do it. Edit: That doesn't do it. Yeah, the padding is still missing in the writing.

pleonex replied:
Yes, the ` fileLength` is padded, so we write the correct length value.
But we are not writing the actual padding bytes (`writer.WritePadding`). So the offset we calculate won't be right.
Edit: you are faster editing than me replying :smile:

WorstAquaPlayer replied:
Just realized that and edited the comment, now it should be padded.

pleonex commented on src/Lemon/Containers/Converters/Binary2Ncch.cs:
```diff
@@ -62,15 +80,27 @@ public Ncch Convert(BinaryFormat source)
 
             // Read the subfiles
             source.Stream.Position = 0x190;
+            header.SdkInfoOffset = reader.ReadInt32();
```
I think we don't need to store the offsets of the files. The problem is that the content of the file changes and it's bigger / smaller, then the offset should also change when we are writing the NCCH (`Ncch2Binary`). So I propose to remove these Offset fields from the header class and in the `Ncch2Binary` re-calculate the offset as we know where are writing each file.

WorstAquaPlayer replied:
This is true, but it was done like this to generate an equal stream to the original one, since the games I tested had some arbitrary positioning, like the sdk_info.txt being written at unit 15 first, and later the logo.bin at unit 5.

pleonex replied:
It could be related to the padding. Maybe if we write the padding we get the same offsets. Can you try?

WorstAquaPlayer replied:
I don't think it's because of the padding, the `WriteOffsetSizeAndData` method now relies on its `offsetInUnits` argument, but originally that method wrote at the end of the file. Changing this like it was, with the padding changes it would pass the tests on games like AA: Dual Destinies, but like I said, there are some games (like PW:AAT) that arbitrarily decides some positions really far in the binary, like the sdk_info.txt that has the offset value of 21 units, but the next file (logo.bin) has the offset value of 5 units.

pleonex replied:
I see... let me check quickly that game. I would like to try to figure out how they decide where to put each file, although I know (like the case of NDS) that some times it's a bit random.

pleonex replied:
You are totally right. I have dumped a couple of games from my 3DS with GodMode9, including PW:AAT, and I can't figure out any logic behind the amount of padding. I'll keep checking.
But for now, instead of adding complexity into the code just to have the same bytes, I propose we do the following:

- Write the files in this order that seem to follow every game: logo, plain, exefs, romfs.
- Change the test to do a "three way" test, so instead of checking the binary bytes, we convert the binary again and we check the new NCCH structure. Something like this:
```csharp
using var ncch = containerConverter.Convert(original);
using var generatedBinary = binaryConverter.Convert(ncch);
using var generatedNcch = containerConverter.Convert(generatedBinary);
ValidateHeader(generatedNcch);
ValidateRegions(generatedNcch);
```

What do you think?

WorstAquaPlayer replied:
Seems like a nice idea. (Edit: Sorry if you saw the message previous to this, I had the wrong order in my .yml and that's why I assumed that)

pleonex replied:
Oh, that's weird. As you can see in this picture (first line is pos 0x190) from my dump of PW:AAT (europe), the order is logo, plain, exefs and romfs.
![image](https://user-images.githubusercontent.com/3107481/148400224-f8f49ded-34eb-4bbb-8e53-5146c9f83331.png)

Then, even if that's different, let's put it in any order. I think _plain, logo, exefs, romfs_ will make the code easier.
By the way, in the _three way_ test don't check the offsets of the files, it won't be right due to the padding mismatches.

WorstAquaPlayer replied:
Yeah, I just edited the comment, I was wrong. That seems to be the correct order. Everything seems ok, so I'll start with the previous corrections and the three-way test.

pleonex reviewed (changes requested):
Thanks a lot! I like the code and implementation!

Just a few comments regarding the padding and offsets as I think we could have some issues after editing the content of the ROM.

pleonex commented:
The build failed because of a known issue with the build system. I'll try to fix it in a different PR quickly.

WorstAquaPlayer committed:
Apply comment corrections and change test behaviour

WorstAquaPlayer committed:
Simplify Ncch2Binary process.
Instead of getting all the info from the original files and forcing to write them at a certain offset, it just does it at the end of the file. Works with the new check.

WorstAquaPlayer committed:
Remove unused offset properties

WorstAquaPlayer committed:
Added missing padding when writing the files

pleonex reviewed (approved):
All good! Thanks a lot
